### ISSUE TITLE
Mxnet Seq2Seq task cost huge memory

### ISSUE THREAD

**Issue by gmyofustc** [apache/incubator-mxnet#3899]
Recently, I use Mxnet for seq2seq task, such as Machine Traslation, however, I find mxnet will cost huge memory, even twice as that of Theano.

I find that the memory cost very strange. 
Eg. 
The softmax layer is (1024, 15w)， I set minibatch=1  Decoder seq_len=40  it will cost 11G memory
If i set softmax layer  (512, 15w),  minibatch=1  Decoder seq_len=40  it cost 8G memory
if i set softmax  layer (256, 15w), minibatch=1  Decoder seq_len=40 it cost 6G memory

Again, if i tune the minibatch
as softmax layer (512, 15w), minibatch=16 Decoder seq_len=40 it cost 9.5G memory , while minibatch=1, it cost 8G memory

In the meanwhile, if I use theano, 
the softmax layer is (1024, 15w), minibatch=32  Decoder seq_len=40 it onlyt cost 4~5G memory!!!!

Not only for machine translation, I've tried lots of other seq2seq tasks,  mxnet always cost huge memory.
As far as i know, so far, mxnet/example has not give a seq2seq example yet, while seq2seq task are so popular for deeplearning, I am really looking foward to hearing advices from guys using mxnet for seq2seq task.

**Comment by feiyulv** [comment]
Mostly caused by the optimizer. For this situation the model size  is huge, and it seems the ndarray calculation in  optimizer  will allocate extra memory for some complex computation. like this:
```
 acc_g[:] = self.rho * acc_g + (1. - self.rho) * grad * grad
```
I'm not familiar with the mechanism of ndarray. Is there any other way to avoid the extra memory allocation except writing a c++ optimizer @piiswrong

**Comment by piiswrong** [comment]
use ccsgd.
If you use the nnvm branch, it's much easier to write an c++ optimizer. You only need to write an op, as shown in nnvm's sgd and adam implementation

**Comment by magic282** [comment]
@gmyofustc Hi, I am also trying this. Did you use attention? And did you compared the training speed with theano?

**Comment by sxjscience** [comment]
@gmyofustc I would be really helpful if you can get some demo codes with dummy data.

**Comment by gmyofustc** [comment]
@magic282 yes, I did use attention, and the speed has no big difference with that of theano

**Comment by gmyofustc** [comment]
@sxjscience sorry for my late reply. Currently, I can not get the code from my inner network environment, but I think I can write the seq2seq code personally in my free time, and contribute to MxNet/Example, perhaps I need one week

**Comment by sxjscience** [comment]
@gmyofustc Thanks a lot! Just take your time.

**Comment by sxjscience** [comment]
Close it now as there is no discussion for a long time.
